packages/core/src/lib/connection_manager.ts

```diff
@@ -0,0 +1,381 @@
+  //  * This is only used when other discovery mechanisms haven't yield peers that are dialable
+  //  * This increases relative centralization of the network
+  //  */
+  // maxTimeBeforeBootstrapFallbackMs?: number;
+  /**
+   * Interval at which to log the state of the connection manager
+   */
+  loggingIntervalMs?: number;
+}
+
+export class ConnectionManager {
+  private static instances = new Map<string, ConnectionManager>();
+  private keepAliveManager = new KeepAliveManager();
+  private options: Options;
+  private libp2pComponents: Libp2pComponents;
+  private dialAttemptsForPeer: Map<string, number> = new Map();
+
+  public static create(
+    peerId: string,
+    libp2p: Libp2pComponents,
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay,
+    options?: Options
+  ): ConnectionManager {
+    let instance = ConnectionManager.instances.get(peerId);
+    if (!instance)
+      instance = new ConnectionManager(
+        libp2p,
+        keepAliveOptions,
+        relay,
+        options
+      );
+    ConnectionManager.instances.set(peerId, instance);
+
+    return instance;
+  }
+
+  constructor(
+    libp2pComponents: Libp2pComponents,
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay,
+    options?: Options
+  ) {
+    this.libp2pComponents = libp2pComponents;
+    this.options = options ?? {};
+
+    this.runService(keepAliveOptions, relay)
+      .then(() => log(`Connection Manager is now running`))
+      .catch((error) => log(`Unexpected error while running service`, error));
+  }
+
+  private async runService(
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay
+  ): Promise<void> {
+    const { loggingIntervalMs = DEFAULT_LOGGING_INTERVAL_MS } = this.options;
+
+    // log state every N seconds
+    setInterval(async () => {
+      this.fetchUpdatedState();
+    }, loggingIntervalMs);
+
+    // start event listeners
+    this.startPeerDiscoveryListener();
+    this.startPeerConnectionListener(keepAliveOptions, relay);
+    this.startPeerDisconnectionListener();
+  }
+
+  stopService(): void {
+    this.keepAliveManager.stopAll();
+    this.libp2pComponents.connectionManager.removeEventListener("peer:connect");
+    this.libp2pComponents.connectionManager.removeEventListener(
+      "peer:disconnect"
+    );
+    this.libp2pComponents.peerStore.removeEventListener("peer");
+  }
+
+  private async dialPeer(peerId: PeerId): Promise<void> {
+    const { maxDialAttemptsForPeer = DEFAULT_MAX_DIAL_ATTEMPTS_FOR_PEER } =
+      this.options;
+
+    try {
+      log(`Dialing peer ${peerId.toString()}`);
+      await this.libp2pComponents.dial(peerId);
+
+      const tags = await this.getTagNamesForPeer(peerId);
+      // add tag to connection describing discovery mechanism
+      this.libp2pComponents.connectionManager
+        .getConnections(peerId)
+        .forEach((conn) => conn.tags.push(...tags));
```

Comment:
`conn` here is an object which we mutate and accumulate `tags` in it without removing duplicates.
Is it possible to have duplicates or end up with having junk in `conn.tags`?